File: app/src/main/java/com/typ/muslim/models/HijriDate.java
/*
 * This app is developed by AHMED SLEEM
 *
 * Copyright (c) 2021.  TYP INC. All Rights Reserved
 */

package com.typ.muslim.models;

import android.content.Context;

import com.typ.muslim.app.AnaMuslimApp;
import com.typ.muslim.managers.HijriCalendar;

import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.util.Objects;

/**
 * Model class representing HijriDate
 */
public class HijriDate implements Serializable {

	/**
	 * Hijri Day of this month
	 */
	private final int day;
	/**
	 * Zero-based Hijri Month index. 0 to 11
	 */
	private final int month;
	/**
	 * Hijri year
	 */
	private final int year;

	public HijriDate(int year, int month, int day) {
		this.day = day;
		this.month = month;
		this.year = year;
	}

	public int getDay() {
		return day;
	}

	public int getMonth() {
		return month;
	}

	public int getYear() {
		return year;
	}

	public String getMonthName() {
		return getMonthName(AnaMuslimApp.getContext().get());
	}

	public String getMonthName(Context context) {
		return HijriCalendar.getHijriMonthName(context, this.month);
	}

	public Timestamp toGregorian() {
		return HijriCalendar.toGregorian(this);
	}

	@Override
	public boolean equals(Object another) {
		if (another == null) return false;
		if (this == another) return true;
		if (!(another instanceof HijriDate)) return false;
		HijriDate hijriDate = (HijriDate) another;
		return day == hijriDate.day &&
		       month == hijriDate.month &&
		       year == hijriDate.year;
	}

	public boolean isAfter(HijriDate another) {
		if (another == null) return false;
		return (day + month + year) - (another.day + another.month + another.year) > 0;
	}

	public boolean isBefore(HijriDate another) {
		if (another == null) return false;
		return (day + month + year) - (another.day + another.month + another.year) < 0;
	}

	@Override
	public int hashCode() {
		return Objects.hash(day, month, year);
	}

	@NotNull
	@Override
	public String toString() {
		return "HijriDate{" +
		       "day=" + day +
		       ", month=(" + getMonthName() + " | " + month +
		       "), year=" + year +
		       '}';
	}


}

File: app/src/main/java/com/typ/muslim/features/calendar/models/HijriDate.kt
/*
 * This app is developed by AHMED SLEEM
 *
 * Copyright (c) 2021.  TYP INC. All Rights Reserved
 */
package com.typ.muslim.features.calendar.models

import android.content.Context
import com.typ.muslim.app.AnaMuslimApp
import com.typ.muslim.features.calendar.HijriCalendar.getHijriMonthName
import com.typ.muslim.features.calendar.HijriCalendar.toGregorian
import com.typ.muslim.models.Timestamp
import java.io.Serializable
import java.util.*

/**
 * Model class representing HijriDate
 */
open class HijriDate(
    val year: Int,
    /**
     * Zero-based Hijri Month index. 0 to 11
     */
    val month: Int,
    /**
     * Hijri Day of this month
     */
    val day: Int
) : Serializable {

    val monthName: String
        get() = getMonthName(AnaMuslimApp.getContext().get())

    fun getMonthName(context: Context?): String {
        return getHijriMonthName(context, month)
    }

    fun toGregorian(): Timestamp? {
        return toGregorian(this)
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (this === other) return true
        if (other !is HijriDate) return false
        return day == other.day && month == other.month && year == other.year
    }

    fun isAfter(another: HijriDate?): Boolean {
        return if (another == null) false else day + month + year - (another.day + another.month + another.year) > 0
    }

    fun isBefore(another: HijriDate?): Boolean {
        return if (another == null) false else day + month + year - (another.day + another.month + another.year) < 0
    }

    override fun hashCode(): Int {
        return Objects.hash(day, month, year)
    }

    override fun toString(): String {
        return "HijriDate{" +
                "day=" + day +
                ", month=(" + monthName + " | " + month +
                "), year=" + year +
                '}'
    }
}